_posts/2015-09-24-mvvm-in-swift.markdown

```diff
@@ -0,0 +1,234 @@
+- collection view layouts.
+- image caching.
+- background-thread data processing. 
+
+It's quite terrifying now that I think about it!
+
+Well, [finally I was ready](https://github.com/artsy/eidolon/pull/503/files). The view controller is now down to 224 lines, and is responsible for only things like displaying data. Calculating _what_ to display and _when_ to display it is now contained within the view model. In true MVVM form, our view controller doesn't even have direct access to the models it displays!
+
+So what does MVVM in Swift look like? Well, our answer is just that – _our_ answer. Others exist, and they have merits and faults of their own. 
+
+I'm not here to preach a definitive definition of MVVM in Swift. Instead, I want to talk about some of the lessons we learnt in the process of building a solution that worked for us.
+
+---
+
+Programming boils down to getting certain bits of code to communicate with other bits in an intelligent way. Software _patterns_ are ways of constraining programmers to make it easier to program.
+
+MVVM, roughly, has the following constraints:
+
+- Models don't talk to anybody (same as MVC).
+- View models only talk to models.
+- View controllers can't talk to models directly; they only interact with view models and views.
+- Views only talk to the view _controllers_, notifying them of interaction events (same as MVC).
+
+And that's pretty much it. It's not that different from MVC – the key differences are:
+
+- There's a new "view model" class. 
+- The view controller no longer has access to the model.
+
+![MVVM Diagram](/images/2015-09-24-mvvm-in-swift/mvvm.png)
+
+Additionally, MVVM on iOS acknowledges the one-to-one relationship between views and view controllers. I tend to think of them as one entity that just happens to be split across a `.swift` file and a Storyboard.
+
+The view model's job is to handle all presentation logic. If a model contains an `NSDate`, the `NSDateFormatter` to format that date would live in the view model. 
+
+View models don't have _any_ access to the user interface. You should not even `import UIKit` in a view model. Typically, a view controller observes the view model somehow to know when there's new data to display. This can be done through KVO or FRP.
+
+MVVM and MVC share a common weakness: neither defines where the network logic of an app should go. I've put it in the view model for now, but I plan on separating it out into its own object soon. That object will be owned by the view model.
+
+---
+
+So let's talk about some specific challenges we had.
+
+## User Interface Structure
+
+Part of our user interface consists of a segment control near the top of the screen. The currently selected segment determines the sort order of the collection view cells, as well as the collection view's layout. We had previously defined an enum to store the titles and sort order corresponding to each segment; the order of the enum cases inferred the order of the buttons in the UI. 
```

Comment:
Small change, "segmented controls" is Apple's nomenclature for this, as opposed to "segment"